server/service/live_queries.go

```diff
@@ -278,9 +278,17 @@ func (svc *Service) RunLiveQueryDeadline(
 			defer cancelFunc()
 
 			defer func() {
+				// We do not want to use the outer `ctx` directly because we want to cleanup the campaign
+				// even if the outer `ctx` is canceled (e.g. a client terminating the connection).
+				ctx := context.WithoutCancel(ctx)
 				err := svc.CompleteCampaign(ctx, campaign)
 				if err != nil {
-					resultsCh <- fleet.QueryCampaignResult{QueryID: queryID, Error: ptr.String(err.Error()), Err: err}
+					level.Error(svc.logger).Log("msg", "completing campaign (sync)", "query.id", campaign.QueryID, "err", err)
+					resultsCh <- fleet.QueryCampaignResult{
```

Comment:
Is it possible that the results Chan has been closed by the time this defer fires? Wouldn't this panic trying to send data to a closed channel?

Reply:
No this is safe, the `close(resultsCh)` call is protected by the `wg` WaitGroup that waits for the goroutine (where this `defer` is) to finish before closing. Since the `defer`s are executed in reverse order, the `wg.Done()` call will be executed last.